src/Raven.Server/Documents/Handlers/Processors/Attachments/AttachmentHandlerProcessorForBulkPostAttachment.cs

```diff
@@ -0,0 +1,126 @@
+using System.Collections.Generic;
+using System.IO;
+using System.Threading.Tasks;
+using JetBrains.Annotations;
+using Raven.Client.Documents.Attachments;
+using Raven.Client.Documents.Operations.Attachments;
+using Raven.Client.Extensions;
+using Raven.Server.Documents.Handlers.Processors.Attachments.Strategies;
+using Raven.Server.Documents.PeriodicBackup.DirectDownload;
+using Raven.Server.ServerWide;
+using Raven.Server.ServerWide.Context;
+using Sparrow.Json;
+
+namespace Raven.Server.Documents.Handlers.Processors.Attachments
+{
+    internal sealed class AttachmentHandlerProcessorForBulkPostAttachment : AbstractAttachmentHandlerProcessorForBulkPostAttachment<DatabaseRequestHandler, DocumentsOperationContext>
+    {
+        public AttachmentHandlerProcessorForBulkPostAttachment([NotNull] DatabaseRequestHandler requestHandler) : base(requestHandler)
+        {
+        }
+
+        protected override async ValueTask GetAttachmentsAsync(DocumentsOperationContext context, BlittableJsonReaderArray attachments, AttachmentType type,
+            OperationCancelToken tcs)
+        {
+            var tasks = new List<Task<Stream>>();
+            var attachmentsStreams = new List<Stream>();
+            DirectFileDownloader downloader = null;
+            bool canDisposeReadTransaction = true;
+            using DocumentsTransaction tx = context.OpenReadTransaction();
+            var stream = new MemoryStream();
+            try
+            {
+                await using (var writer = new AsyncBlittableJsonTextWriter(context, stream))
+                {
+                    writer.WriteStartObject();
+                    writer.WritePropertyName(nameof(GetAttachmentsOperation.GetAttachmentsCommand.AttachmentsMetadata));
+                    writer.WriteStartArray();
+                    var first = true;
+
+                    foreach (BlittableJsonReaderObject bjro in attachments)
+                    {
+                        if (bjro.TryGet(nameof(AttachmentRequest.DocumentId), out string id) == false)
+                            throw new ArgumentException($"Could not parse {nameof(AttachmentRequest.DocumentId)}");
+                        if (bjro.TryGet(nameof(AttachmentRequest.Name), out string name) == false)
+                            throw new ArgumentException($"Could not parse {nameof(AttachmentRequest.Name)}");
+
+                        var attachment = RequestHandler.Database.DocumentsStorage.AttachmentsStorage.GetAttachment(context, id, name, type, changeVector: null);
+                        if (attachment == null)
+                            continue;
+
+                        IBulkPostAttachmentStrategy strategy;
+                        if (attachment.RemoteParameters.IsRemoteStorageAttachment())
+                        {
+                            strategy = new RemoteBulkPostAttachmentStrategyProcessor(RequestHandler);
+                            downloader ??= strategy.GetAttachmentsDownloader(attachment, tcs);
```

Comment:
I see you get attachments stream in parallel and dispose the `downloader` once at the end, can the `downloader` be different instance from one attachment to another?